src/Dialect/ONNX/ONNXDimAnalysis.cpp

```diff
@@ -369,176 +500,122 @@ void DimAnalysis::visitDim(
+  exploreSameDimsUsingShapeHelper(dim, op, sameDims);
+
+  ////////////////////////////////////////////////////
+  // For operations that have an input specifying the output shape, the output
+  // dim can come from the shape input.
+  // For example: ConstantOfShape, Expand, Reshape.
+  exploreSameDimsUsingShapeInput(dim, op, sameDims);
 
-  // MatMulOp
-  if (auto matmulOp = dyn_cast<ONNXMatMulOp>(op)) {
-    bool success = exploreSameInputDims(dim, op, sameDims);
-    assert(success && "dim analysis should be successful for matmul ops ");
-
-    // If we know by this analysis that two dynamic dims at the same index in
-    // the batchsize space are the same, then the output dim must be the same
-    // too.
-    Value A = matmulOp.getA();
-    Value B = matmulOp.getB();
-    Type aType = A.getType();
-    Type bType = B.getType();
-    uint64_t aRank = getRank(aType);
-    uint64_t bRank = getRank(bType);
-    uint64_t maxRank = std::max(aRank, bRank);
-    if (dimIndex <= maxRank - 2) { // In the batchsize space.
-      ArrayRef<int64_t> aShape = getShape(aType);
-      ArrayRef<int64_t> bShape = getShape(bType);
-      // aDim == bDim (dynamic), there is no broadcasting and aDim ==
-      // outputDim.
-      int64_t aDimIndex = dimIndex - (maxRank - aRank);
-      int64_t bDimIndex = dimIndex - (maxRank - bRank);
-      if ((aDimIndex >= 0) && (bDimIndex >= 0) &&
-          ShapedType::isDynamic(aShape[aDimIndex]) &&
-          ShapedType::isDynamic(bShape[bDimIndex]) &&
-          sameDynDim(A, aDimIndex, B, bDimIndex))
-        sameDims.insert(DimT(A, aDimIndex));
+  ////////////////////////////////////////////////////
+  // Special/additional cases.
+
+  // Variadic/non-variadic BinaryOp (alphabetical list, same as under
+  // ONNXBroadcastOpShapeHelper).
+  // MatMul uses the same broadcasting rule as BinaryOp, include it here. If we
+  // know by this analysis that two dynamic dims at the same index (counting
+  // from the innermost dim back) are the same, then the output dim must be the
+  // same too.
+  if (isa<ONNXAddOp, ONNXAndOp, ONNXBitShiftOp, ONNXBitwiseAndOp,
+          ONNXBitwiseOrOp, ONNXBitwiseXorOp, ONNXBitShiftOp, ONNXDivOp,
+          ONNXEqualOp, ONNXGreaterOp, ONNXGreaterOrEqualOp, ONNXLessOp,
+          ONNXLessOrEqualOp, ONNXMatMulOp, ONNXMeanOp, ONNXMinOp, ONNXModOp,
+          ONNXMulOp, ONNXOrOp, ONNXPowOp, ONNXSubOp, ONNXSumOp, ONNXWhereOp,
+          ONNXXorOp>(op)) {
+    LLVM_DEBUG(llvm::dbgs() << "Explore same inputs\n");
+    OperandRange operands = op->getOperands();
+    for (size_t i = 0; i < operands.size() - 1; ++i) {
+      Value A = op->getOperands()[i];
+      for (size_t k = i + 1; k < operands.size(); ++k) {
+        Value B = op->getOperands()[k];
+        uint64_t aRank = getRank(A.getType());
+        uint64_t bRank = getRank(B.getType());
+        if ((aRank != 0) && (bRank != 0)) {
+          // aDim == bDim (dynamic), there is no broadcasting and aDim ==
+          // outputDim.
```

Comment:
Does the comment means `If aDim == bDim (dynamic), outputDim = aDim. There is no broadcast for this dimension`.

Reply:
Yes. Thanks